### PR TITLE
[WIP]Enable zipcode search for foreign zipcodes, and remove input of 5-digits restricition

### PR DESCRIPTION
**Pull request by hcaofec**
## Summary (required)

- Resolves [#3054](https://github.com/fecgov/openFEC/issues/3054)
1) Enable zipcode search for zipcodes greater than 5 digits.
2) Enable search for foreign zipcode that has letters, space, and -,  and to exclude certain special character.  
3) The results still match on the first 5 characters.
 4) refactor test case accordingly.

## How to test the changes locally
1) run this branch locally, using DEV database
2) test this endpoint locally :  /schedules/schedule_a/
    
    two_year_transaction_period=2016
    contributor_zip: using the value in **zip code** column
    
    

zip code | counts before change | counts after change |  
---- | -- | -- | --
10050 | 26 | 26 |  
100502814 | error code = 400 | 26 |  
10050-2814|error_code = 400 | 26
L2A2W4 | error code = 400 | 3 |  L2A2W4 ( 3 )
L5K 2B3 | error code = 400 | 3 | L5K 2G1;   L5K 2G1;    L5K 2B3

The counts  concur with the results by running query against database directly.

### CODE REVIEW

File: tests/test_itemized.py [hunk 1]
```diff
@@ -111,24 +111,27 @@ def test_filter_sched_a_zip(self):
...
+        results = self._results(api.url_for(ScheduleAView, contributor_zip=967893405, **self.kwargs))
         self.assertEqual(len(results), 3)
     
-    def test_invalid_sched_a_zip(self): 
-        response = self.app.get(api.url_for(ScheduleAView,contributor_zip=9678912,cycle=2018))
-        self.assertEqual(response.status_code,400)
+        results = self._results(api.url_for(ScheduleAView, contributor_zip='M4C 1M55', **self.kwargs))
+        self.assertEqual(len(results), 1)
+
+        contributor_zips = ['M4C 1M5555',96789]
```
Comment (vrajmohan) [inline tests/test_itemized.py]: PEP8 violation

File: tests/test_itemized.py [hunk 2]
```diff
@@ -111,24 +111,27 @@ def test_filter_sched_a_zip(self):
...
-
-        results = self._results(api.url_for(ScheduleAView, contributor_zip=96789, **self.kwargs))
+        results = self._results(api.url_for(ScheduleAView, contributor_zip=967893405, **self.kwargs))
         self.assertEqual(len(results), 3)
     
-    def test_invalid_sched_a_zip(self): 
-        response = self.app.get(api.url_for(ScheduleAView,contributor_zip=9678912,cycle=2018))
-        self.assertEqual(response.status_code,400)
+        results = self._results(api.url_for(ScheduleAView, contributor_zip='M4C 1M55', **self.kwargs))
+        self.assertEqual(len(results), 1)
```
Comment (vrajmohan) [inline tests/test_itemized.py]: Are we sure this is intended? 'M4C 1M55' will match 'M4C 1M7'.

File: webservices/resources/sched_a.py [hunk 3]
```diff
@@ -80,16 +80,18 @@ def args(self):
     def build_query(self, **kwargs):
         query = super().build_query(**kwargs)
         query = filters.filter_contributor_type(query, self.model.entity_type, kwargs)
-
+        zip_list=[]
         if kwargs.get('contributor_zip'):
             for value in kwargs['contributor_zip']:
-                if re.search('^-?\d{5}$',value) is None:
+                if re.search('[^a-zA-Z0-9-\s]+',value):
```
Comment (vrajmohan) [inline webservices/resources/sched_a.py]: We don't need the trailing plus.

**Comment by hcaofec** [comment]
are we using the same matching on first 5 digits/characters rule for foreign zipcodes?  
for example, if searing for 'L5K 2B3',  the below three rows will return with satisfying condition,
zipcode like 'L5K 2%'

 L5K 2G1;   L5K 2G1;    L5K 2B3

@PaulClark2 @jwchumley @lbeaufort

**Comment by PaulClark2** [comment]
@hcaofec I think matching on the first five characters is an okay place to start. I do have one question. What happens if the postal code is less than five characters? For example, Iceland uses 3-digit postal codes.

**Comment by vrajmohan** [comment]
My $0.02: If we detect the zipcode to be 9-digit US (with or without a - between 5th and 6th), use the 1st 5 digits for a match. In all other cases, perform an exact match.

**Comment by hcaofec** [comment]
If API searches by 10050, then zipcode=100502814 in database will not be returned. This is listed as one of the issues to be fixed in original ticket.

@vrajmohan

**Comment by vrajmohan** [comment]
I was thinking that when we store the postal code we would create an expression based index (which uses the above logic) and then query it using the same expression.

File: tests/test_itemized.py (hunk 2 above)
Comment (hcaofec) [reply tests/test_itemized.py]: yes for now since we would rather be over-inclusive than under-inclusive

**Comment by lbeaufort** [comment]
@hcaofec I'm sorry I ran out of time to finish reviewing this but will look on Monday. I'm having an issue with testing some London zipcodes, and want to find more test examples in the data. 

For this example, it finds no results but shows `count: 8412, pages: 281`

http://127.0.0.1:5000/v1/schedules/schedule_a/?api_key=DEMO_KEY&sort_hide_null=true&data_type=processed&two_year_transaction_period=2018&contributor_zip=WC2N+5DU&contributor_zip=SW1A+1AA&sort=-contribution_receipt_date&per_page=30

screen shot 2018-07-20 at 3 47 54 pm

**Comment by hcaofec** [comment]
looking into it. @lbeaufort

**Comment by hcaofec** [comment]
One solution is using actual count of the query like [this](https://github.com/fecgov/openFEC/blob/develop/webservices/common/counts.py#L20)

But in current api, since there is no minimum requirement of the length of search string, nor there is maximum restriction on how many zipcodes can be searched, the result set could be large if user searches by 1 digit or 5+ zipcodes at once, which will be very slow to get count.

zipcode | actual count
-- | --
22 | 884748
20170,22101,221 | 144415
20170,22101|35501 

In order to get accurate count in time, we may want to reduce the size of results by endpoint by adding this:
1) set minimum length of zipcode = 3 ( this is the same rule used in classic website ). For example, 2, 23 are not accepted. 123, 1234, 12345, 123465 are ok
2) disable multiple zipcode search. Only 1 zipcode per api call.

Another possible solution is refactor the query to have planner return more accurate estimate.

**Comment by JonellaCulmer** [comment]
@hcaofec Are we good to go on this with the review from @vrajmohan, or does @lbeaufort still need to review before we merge this PR?

**Comment by lbeaufort** [comment]
I've confirmed the count issue is an existing one, and shouldn't hold up the foreign zip code work. I'm going to keep testing today. Example: https://api.open.fec.gov/v1/schedules/schedule_a/?api_key=DEMO_KEY&sort_hide_null=true&data_type=processed&two_year_transaction_period=2018&contributor_zip=88888&contributor_zip=77777&sort=-contribution_receipt_date&per_page=30

**Pull request merged**